File: apps/web/src/lib/api/coaching.ts
/**
 * Coaching API
 * API client for coaching sessions and packages
 */

import { apiClient } from './client';

export interface CoachingPackage {
  id: number;
  name: string;
  description: string | null;
  price: number;
  currency: string;
  duration_months: number | null;
  sessions_count: number;
  features: string | null;
  is_active: boolean;
  is_popular: boolean;
  stripe_price_id: string | null;
  stripe_product_id: string | null;
  created_at: string;
  updated_at: string;
}

export interface Coach {
  id: number;
  name: string;
  email: string;
  first_name: string | null;
  last_name: string | null;
  avatar: string | null;
}

export interface CoachingSession {
  id: number;
  user_id: number;
  coach_id: number;
  package_id: number | null;
  scheduled_at: string;
  duration_minutes: number;
  status: string;
  amount: number;
  currency: string;
  stripe_checkout_session_id: string | null;
  stripe_payment_intent_id: string | null;
  payment_status: string | null;
  notes: string | null;
  coach_notes: string | null;
  created_at: string;
  updated_at: string;
}

export interface CreateSessionData {
  coach_id: number;
  package_id?: number | null;
  scheduled_at: string;
  duration_minutes?: number;
  notes?: string | null;
}

export interface CheckoutSessionData {
  session_id: number;
  success_url: string;
  cancel_url: string;
}

export interface CheckoutSessionResponse {
  session_id: string;
  url: string;
}

export interface TimeSlot {
  start: string;
  end: string;
  available: boolean;
}

export interface CoachAvailabilityRequest {
  coach_id: number;
  start_date: string;
  end_date: string;
}

export interface CoachAvailabilityResponse {
  coach_id: number;
  slots: TimeSlot[];
}

/**
 * Get all coaching packages
 */
export async function getCoachingPackages(activeOnly: boolean = true): Promise<CoachingPackage[]> {
  const response = await apiClient.get<{ packages: CoachingPackage[]; total: number }>(
    `/v1/coaching/packages?active_only=${activeOnly}`
  );
  return response.data?.packages || [];
}

/**
 * Get coaching package by ID
 */
export async function getCoachingPackage(packageId: number): Promise<CoachingPackage> {
  const response = await apiClient.get<CoachingPackage>(`/v1/coaching/packages/${packageId}`);
  return response.data!;
}

/**
 * Get all coaches
 */
export async function getCoaches(): Promise<Coach[]> {
  const response = await apiClient.get<Coach[]>(`/v1/coaching/coaches`);
  return response.data || [];
}

/**
 * Get coach by ID
 */
export async function getCoach(coachId: number): Promise<Coach> {
  const response = await apiClient.get<Coach>(`/v1/coaching/coaches/${coachId}`);
  return response.data!;
}

/**
 * Create a coaching session
 */
export async function createCoachingSession(data: CreateSessionData): Promise<CoachingSession> {
  const response = await apiClient.post<CoachingSession>(`/v1/coaching/sessions`, data);
  return response.data!;
}

/**
 * Get user's coaching sessions
 */
export async function getMyCoachingSessions(
  asCoach: boolean = false,
  statusFilter?: string
): Promise<CoachingSession[]> {
  const params = new URLSearchParams();
  if (asCoach) params.append('as_coach', 'true');
  if (statusFilter) params.append('status_filter', statusFilter);
  
  const response = await apiClient.get<{ sessions: CoachingSession[]; total: number }>(
    `/v1/coaching/sessions/me?${params.toString()}`
  );
  return response.data?.sessions || [];
}

/**
 * Get coaching session by ID
 */
export async function getCoachingSession(sessionId: number): Promise<CoachingSession> {
  const response = await apiClient.get<CoachingSession>(`/v1/coaching/sessions/${sessionId}`);
  return response.data!;
}

/**
 * Create Stripe checkout session for a coaching session
 */
export async function createCoachingCheckoutSession(
  sessionId: number,
  data: CheckoutSessionData
): Promise<CheckoutSessionResponse> {
  const response = await apiClient.post<CheckoutSessionResponse>(
    `/v1/coaching/sessions/${sessionId}/checkout`,
    data
  );
  return response.data!;
}

/**
 * Get coach availability
 */
export async function getCoachAvailability(
  coachId: number,
  startDate: string,
  endDate: string
): Promise<CoachAvailabilityResponse> {
  const response = await apiClient.post<CoachAvailabilityResponse>(
    `/v1/coaching/coaches/${coachId}/availability`,
    {
      coach_id: coachId,
      start_date: startDate,
      end_date: endDate,
    }
  );
  return response.data!;
}

File: backend/app/api/v1/endpoints/coaching.py
"""
Coaching Endpoints
API endpoints for coaching session and package management
"""

from typing import List
from datetime import datetime, timedelta
from fastapi import APIRouter, Depends, HTTPException, status, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.database import get_db
from app.dependencies import get_current_user, get_stripe_service
from app.models import User
from app.services.coaching_service import CoachingService
from app.services.stripe_service import StripeService
from app.schemas.coaching import (
    CoachingPackageResponse,
    CoachingPackageListResponse,
    CoachingSessionCreate,
    CoachingSessionResponse,
    CoachingSessionListResponse,
    CheckoutSessionCreate,
    CheckoutSessionResponse,
    CoachAvailabilityRequest,
    CoachAvailabilityResponse,
    TimeSlot,
)

router = APIRouter(prefix="/coaching", tags=["coaching"])


def get_coaching_service(db: AsyncSession = Depends(get_db)) -> CoachingService:
    """Dependency to get coaching service"""
    return CoachingService(db)


@router.get("/packages", response_model=CoachingPackageListResponse)
async def list_packages(
    active_only: bool = Query(True, description="Only return active packages"),
    coaching_service: CoachingService = Depends(get_coaching_service),
):
    """List all available coaching packages"""
    packages = await coaching_service.get_all_packages(active_only=active_only)
    
    return CoachingPackageListResponse(
        packages=[CoachingPackageResponse.model_validate(pkg) for pkg in packages],
        total=len(packages)
    )


@router.get("/packages/{package_id}", response_model=CoachingPackageResponse)
async def get_package(
    package_id: int,
    coaching_service: CoachingService = Depends(get_coaching_service),
):
    """Get coaching package by ID"""
    package = await coaching_service.get_package(package_id)
    
    if not package:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Package not found"
        )
    
    return CoachingPackageResponse.model_validate(package)


@router.get("/coaches", response_model=List[dict])
async def list_coaches(
    coaching_service: CoachingService = Depends(get_coaching_service),
):
    """List all available coaches"""
    coaches = await coaching_service.get_all_coaches()
    
    return [
        {
            "id": coach.id,
            "name": f"{coach.first_name} {coach.last_name}".strip() or coach.email,
            "email": coach.email,
            "first_name": coach.first_name,
            "last_name": coach.last_name,
            "avatar": coach.avatar,
        }
        for coach in coaches
    ]


@router.get("/coaches/{coach_id}", response_model=dict)
async def get_coach(
    coach_id: int,
    coaching_service: CoachingService = Depends(get_coaching_service),
):
    """Get coach by ID"""
    coach = await coaching_service.get_coach(coach_id)
    
    if not coach:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Coach not found"
        )
    
    return {
        "id": coach.id,
        "name": f"{coach.first_name} {coach.last_name}".strip() or coach.email,
        "email": coach.email,
        "first_name": coach.first_name,
        "last_name": coach.last_name,
        "avatar": coach.avatar,
    }


@router.post("/sessions", response_model=CoachingSessionResponse, status_code=status.HTTP_201_CREATED)
async def create_session(
    session_data: CoachingSessionCreate,
    current_user: User = Depends(get_current_user),
    coaching_service: CoachingService = Depends(get_coaching_service),
):
    """Create a new coaching session"""
    try:
        session = await coaching_service.create_session(
            user_id=current_user.id,
            coach_id=session_data.coach_id,
            scheduled_at=session_data.scheduled_at,
            duration_minutes=session_data.duration_minutes,
            package_id=session_data.package_id,
            notes=session_data.notes,
        )
        
        return CoachingSessionResponse.model_validate(session)
    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(e)
        )


@router.get("/sessions/me", response_model=CoachingSessionListResponse)
async def get_my_sessions(
    as_coach: bool = Query(False, description="Get sessions as coach"),
    status_filter: str = Query(None, description="Filter by status"),
    current_user: User = Depends(get_current_user),
    coaching_service: CoachingService = Depends(get_coaching_service),
):
    """Get current user's coaching sessions"""
    from app.models.coaching_session import SessionStatus
    
    status_enum = None
    if status_filter:
        try:
            status_enum = SessionStatus(status_filter)
        except ValueError:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail=f"Invalid status: {status_filter}"
            )
    
    sessions = await coaching_service.get_user_sessions(
        user_id=current_user.id,
        as_coach=as_coach,
        status=status_enum
    )
    
    return CoachingSessionListResponse(
        sessions=[CoachingSessionResponse.model_validate(s) for s in sessions],
        total=len(sessions)
    )


@router.get("/sessions/{session_id}", response_model=CoachingSessionResponse)
async def get_session(
    session_id: int,
    current_user: User = Depends(get_current_user),
    coaching_service: CoachingService = Depends(get_coaching_service),
):
    """Get coaching session by ID"""
    session = await coaching_service.get_session(session_id, user_id=current_user.id)
    
    if not session:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Session not found"
        )
    
    return CoachingSessionResponse.model_validate(session)


@router.post("/sessions/{session_id}/checkout", response_model=CheckoutSessionResponse)
async def create_checkout_session(
    session_id: int,
    checkout_data: CheckoutSessionCreate,
    current_user: User = Depends(get_current_user),
    coaching_service: CoachingService = Depends(get_coaching_service),
    stripe_service: StripeService = Depends(get_stripe_service),
):
    """Create Stripe checkout session for a coaching session"""
    # Verify session belongs to user
    session = await coaching_service.get_session(session_id, user_id=current_user.id)
    
    if not session:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Session not found"
        )
    
    if session.user_id != current_user.id:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="You don't have permission to access this session"
        )
    
    from app.models.coaching_session import SessionStatus
    if session.status != SessionStatus.PENDING:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Session is not in pending status"
        )
    
    # Create Stripe checkout session
    stripe_session = await stripe_service.create_coaching_checkout_session(
        user=current_user,
        amount=float(session.amount),
        currency=session.currency or "cad",
        session_id=session_id,
        success_url=checkout_data.success_url,
        cancel_url=checkout_data.cancel_url,
        metadata={
            "coach_id": str(session.coach_id),
            "scheduled_at": session.scheduled_at.isoformat(),
        }
    )
    
    # Update session with Stripe checkout session ID
    await coaching_service.update_session_payment(
        session_id=session_id,
        stripe_checkout_session_id=stripe_session["session_id"],
        payment_status="pending"
    )
    
    return CheckoutSessionResponse(**stripe_session)


@router.post("/coaches/{coach_id}/availability", response_model=CoachAvailabilityResponse)
async def get_coach_availability(
    coach_id: int,
    request: CoachAvailabilityRequest,
    coaching_service: CoachingService = Depends(get_coaching_service),
):
    """Get available time slots for a coach"""
    if request.coach_id != coach_id:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Coach ID mismatch"
        )
    
    slots = await coaching_service.get_coach_availability(
        coach_id=coach_id,
        start_date=request.start_date,
        end_date=request.end_date
    )
    
    return CoachAvailabilityResponse(
        coach_id=coach_id,
        slots=[
            TimeSlot(
                start=slot,
                end=slot + timedelta(minutes=60),
                available=True
            )
            for slot in slots
        ]
    )
